src/lowering/desugar.ml

```diff
@@ -513,9 +513,9 @@ and build_actor at ts self_id es obj_typ =
                ) fields vs)
             ty)) in
   let footprint_d, footprint_f = export_footprint self_id (with_stable_vars (fun e -> e)) in
-  I.(ActorE (footprint_d @ ds', footprint_f @ fs,
+  (ty, I.(ActorE (footprint_d @ ds', footprint_f @ fs,
      { meta;
-       preupgrade = with_stable_vars (fun e -> primE (I.ICStableWrite ty) [e]);
+       preupgrade = (primE (I.ICStableWrite ty) []);
```

Comment:
Curious what happened to the argument `e` - I guess the semantics of ICStableWrite has changed a bit...

Reply:
Yes, I had to change the semantics of ICStableWrite. The stable actor record to persist can no longer be passed in as argument (`e`) because explicit incremental stabilization may start before the actual upgrade and already needs the to build the stable record (and run preupgrade) in advance. Therefore, the persistence logic builds the stable record when needed (inversion of control). `ICStableWrite` now means "complete the incremental stabilization" or "prepare EOP upgrade".